types/wlr_output.c

```diff
@@ -188,94 +188,174 @@ void wlr_output_set_position(struct wlr_output *output, int32_t lx,
+	wl_signal_init(&output->events.resolution);
+	wl_signal_init(&output->events.destroy);
+}
+
+void wlr_output_destroy(struct wlr_output *output) {
+	if (!output) {
+		return;
 	}
 
-	wlr_log(L_INFO, "Falling back to software cursor");
+	wl_signal_emit(&output->events.destroy, output);
 
-	output->cursor.is_sw = true;
-	output->cursor.width = width;
-	output->cursor.height = height;
+	struct wlr_output_mode *mode, *tmp_mode;
+	wl_list_for_each_safe(mode, tmp_mode, &output->modes, link) {
+		free(mode);
+	}
+	wl_list_remove(&output->modes);
+	if (output->impl && output->impl->destroy) {
+		output->impl->destroy(output);
+	} else {
+		free(output);
+	}
+}
 
-	if (!output->cursor.renderer) {
-		output->cursor.renderer = wlr_gles2_renderer_create(output->backend);
-		if (!output->cursor.renderer) {
-			return false;
-		}
+void wlr_output_effective_resolution(struct wlr_output *output,
+		int *width, int *height) {
+	// TODO: Scale factor
+	if (output->transform % 2 == 1) {
+		*width = output->height;
+		*height = output->width;
+	} else {
+		*width = output->width;
+		*height = output->height;
 	}
+}
 
-	if (!output->cursor.texture) {
-		output->cursor.texture =
-			wlr_render_texture_create(output->cursor.renderer);
-		if (!output->cursor.texture) {
-			return false;
+void wlr_output_make_current(struct wlr_output *output) {
+	output->impl->make_current(output);
+}
+
+static void output_cursor_render(struct wlr_output_cursor *cursor) {
+	struct wlr_texture *texture = cursor->texture;
+	struct wlr_renderer *renderer = cursor->renderer;
+	if (cursor->surface != NULL) {
+		// Some clients commit a cursor surface with a NULL buffer to hide it.
+		if (!wlr_surface_has_buffer(cursor->surface)) {
+			return;
 		}
+		texture = cursor->surface->texture;
+		renderer = cursor->surface->renderer;
 	}
 
-	return wlr_texture_upload_pixels(output->cursor.texture,
-		WL_SHM_FORMAT_ARGB8888, stride, width, height, buf);
+	if (texture != NULL && renderer != NULL) {
+		glViewport(0, 0, cursor->output->width, cursor->output->height);
+		glEnable(GL_BLEND);
+		glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+
+		float matrix[16];
+		wlr_texture_get_matrix(texture, &matrix,
+			&cursor->output->transform_matrix, cursor->x - cursor->hotspot_x,
+			cursor->y - cursor->hotspot_y);
+		wlr_render_with_matrix(renderer, texture, &matrix);
+	}
 }
 
-bool wlr_output_set_cursor(struct wlr_output *output,
-		const uint8_t *buf, int32_t stride, uint32_t width, uint32_t height,
-		int32_t hotspot_x, int32_t hotspot_y) {
-	if (output->cursor.surface) {
-		wl_list_remove(&output->cursor.surface_commit.link);
-		wl_list_remove(&output->cursor.surface_destroy.link);
-		output->cursor.surface = NULL;
+void wlr_output_swap_buffers(struct wlr_output *output) {
+	struct wlr_output_cursor *cursor;
+	wl_list_for_each(cursor, &output->cursors, link) {
+		if (output->hardware_cursor == cursor) {
+			continue;
+		}
+		output_cursor_render(cursor);
```

Comment:
Is it a problem that this rendering happens in swap buffers and not the frame callback like the rest of the rendering?

Reply:
What if the compositor wants to customize the rendering of the cursor or hide it?

Reply:
A problem is that the `frame` callback is emitted by the output implementation (so by the backend), so we can't just add a listener in `wlr_output` (otherwise we have no guarantee that the cursor drawn will be on top of everything).

(Also this is how it worked before this PR)

@ascent12 Thoughts?

Reply:
>What if the compositor wants to customize the rendering of the cursor or hide it?

The compositor is responsible for creating `wlr_output_cursor`s. So if it doesn't want to render a cursor, it can just do nothing. If it wants to customize how the cursor looks, it can just create a `wlr_output_cursor` with whatever pixels it wants. If it wants to render cursors differently, it can never use `wlr_output_cursor`s and render them itself.